### PR TITLE
add support for Team Reactions API

### LINKED ISSUE #877: Support preview Reactions API
GitHub Developer API announcement:
https://developer.github.com/changes/2018-03-16-team-discussion-reactions-api-and-apps-endpoints/

This would be a great PR for any new contributor to this repo or a new Go developer.
All contributions are greatly appreciated!

Feel free to volunteer for any issue, and we can send you an invite to contribute to the repo (which you then accept after you enable two-factor authentication) and the issue can be assigned to you so that others don't attempt to duplicate the work.

Please check out our [CONTRIBUTING.md](https://github.com/google/go-github/blob/master/CONTRIBUTING.md) guide to get started.

Thank you!

### PR DESCRIPTION
**Pull request by urohit011**
Closes #877.

### CODE REVIEW

**Comment by gmlewis** [comment]
Hi @urohit011, thank you for the PR.

It looks like there are a few issues that need to be resolved before I start a code review.

Please make sure you take a look at our `CONTRIBUTING.md` file... it has some important steps for you to run, like making sure the tests pass locally:

```bash
go generate github.com/google/go-github/...
go test github.com/google/go-github/...
go vet github.com/google/go-github/...
```

Also, here is some output from TravisCI to investigate:

```
# github.com/google/go-github/github
github/reactions.go:274:55: undefined: mediaTypeGraphQLNodeIDPreview
github/reactions.go:275:27: undefined: strings
github/reactions.go:298:55: undefined: mediaTypeGraphQLNodeIDPreview
github/reactions.go:299:27: undefined: strings
github/reactions.go:325:55: undefined: mediaTypeGraphQLNodeIDPreview
github/reactions.go:326:27: undefined: strings
github/reactions.go:346:55: undefined: mediaTypeGraphQLNodeIDPreview
github/reactions.go:347:27: undefined: strings
github/reactions_test.go:191:55: undefined: mediaTypeGraphQLNodeIDPreview
github/reactions_test.go:194:30: undefined: strings
github/reactions_test.go:194:30: too many errors
FAIL	github.com/google/go-github/github [build failed]
```

**Comment by urohit011** [comment]
Got it @gmlewis, I'll do the needful.

**Comment by urohit011** [comment]
Hello @gmlewis, here is what I tried and perhaps I am doing it the wrong way, I ran the following commands that are mentioned in the `travis.yml` file:

`go test -v -race ./...`
`go test -v -tags=integration -run=^$ ./test/integration`

Both these commands run fine on my local machine, but the Travis build is still failing. Could you please help me understand why this is happening?

**Comment by gmlewis** [comment]
Sure, @urohit011.

The first command that fails in TravisCI is this:

> The command "go get -t -v ./..." exited with 2.
> $ diff -u <(echo -n) <(gofmt -d -s .)

This means that it found files where `gofmt` was not run. I've personally integrated `goimports` with my editor so that on every save of every Go file, it runs `goimports` (which has the nice side effect of running `gofmt` to make sure my code has been formatted properly). I highly recommend `goimports` integration with your editor.

Then the next command that fails in TravisCI is this:

> $ go test -v -race ./...
> github/reactions.go:274:55: undefined: mediaTypeGraphQLNodeIDPreview
> github/reactions.go:275:27: undefined: strings

Searching for `mediaTypeGraphQLNodeIDPreview`, I see that it has recently been removed from this repo... so it looks like you need to sync your branch with the main repo. There are help articles on GitHub how to do this, but basically, here is what I usually do:

```bash
$ git checkout master
$ git pull upstream master
$ git checkout my-development-branch
$ git merge master
$ git mergetool    # if there were merge conflicts
$ git commit -am 'Merge master into my-development-branch'
$ git push origin my-development branch
```

Please let me know if this is not clear or you need more help syncing up.
When you `git push` your update, please do not use `--force` so that the reviews are easier for the reviewer.

**Comment by urohit011** [comment]
Thank you very much for the clarification @gmlewis, It looks clear so far, I'll fix it and let you know if you know if I face problems.

**Comment by urohit011** [comment]
Hello @gmlewis, the Travis build passed (except for the test that is allowed to fail), can we go ahead with the code review now?

File: github/reactions.go [hunk 1]
```diff
@@ -256,6 +256,101 @@ func (s ReactionsService) CreatePullRequestCommentReaction(ctx context.Context,
 	return m, resp, nil
 }
 
+// ListTeamDiscussionReactions lists the reactions for a team discussion.
+//
+// GitHub API docs: https://developer.github.com/v3/reactions/#list-reactions-for-a-team-discussion
+func (s *ReactionsService) ListTeamDiscussionReactions(ctx context.Context, teamID, discussionNumber int64, opt *ListOptions) ([]*Reaction, *Response, error) {
```
Comment (gmlewis) [inline github/reactions.go]: I'm thinking that `teamID` (and all IDs on GitHub should remain an `int64`) but that `discussionNumber` and `commentNumber` (reading through the GitHub Developer API docs... it doesn't show up here but probably does below) should be simply `int`.

I think having the discussion and comment number be `int` clues the user of this package into the fact that these are typically small indices and are never expected to be even in the hundreds. I've seen a few GitHub issues go viral before and even then I don't think the indices grew more than a few thousand.

I'd like to hear what @juliaferraioli thinks about this before we make a change, though, so hold off on this until we hear from her.

File: github/reactions.go [hunk 2]
```diff
@@ -256,6 +256,101 @@ func (s ReactionsService) CreatePullRequestCommentReaction(ctx context.Context,
...
+		return nil, resp, err
+	}
+
+	return m, resp, nil
+}
+
+// CreateTeamDiscussionReaction creates a reaction for a team discussion.
+//
+// GitHub API docs: https://developer.github.com/v3/reactions/#create-reaction-for-a-team-discussion
+func (s *ReactionsService) CreateTeamDiscussionReaction(ctx context.Context, teamID, discussionNumber int64, content string) (*Reaction, *Response, error) {
```
Comment (gmlewis) [inline github/reactions.go]: Can you please enumerate the possible values of `content` in the comments above, taken from:
https://developer.github.com/v3/reactions/#reaction-types
?

File: github/reactions.go [hunk 3]
```diff
@@ -256,6 +256,101 @@ func (s ReactionsService) CreatePullRequestCommentReaction(ctx context.Context,
...
+	m := &Reaction{}
+	resp, err := s.client.Do(ctx, req, m)
+	if err != nil {
+		return nil, resp, err
+	}
+
+	return m, resp, nil
+}
+
+//ListTeamDiscussionCommentReactions lists the reactions for a team discussion comment.
```
Comment (gmlewis) [inline github/reactions.go]: In all the comments, please add a single space between the `//` and the first character (e.g. `// ListTeam...`) for consistency with the rest of the repo.

File: github/reactions.go [hunk 4]
```diff
@@ -256,6 +256,101 @@ func (s ReactionsService) CreatePullRequestCommentReaction(ctx context.Context,
...
+
+	var m []*Reaction
+	resp, err := s.client.Do(ctx, req, &m)
+
+	return m, resp, nil
+}
+
+// CreateTeamDiscussionCommentReaction creates a reaction for a team discussion comment
+//
+// GitHub API docs: https://developer.github.com/v3/reactions/#create-reaction-for-a-team-discussion-comment
```
Comment (gmlewis) [inline github/reactions.go]: Please also enumerate the possible values of `content` here as well.

File: github/reactions.go [hunk 5]
```diff
@@ -256,6 +256,101 @@ func (s ReactionsService) CreatePullRequestCommentReaction(ctx context.Context,
...
+
+	req.Header.Set("Accept", mediaTypeReactionsPreview)
+
+	var m []*Reaction
+	resp, err := s.client.Do(ctx, req, &m)
+
+	return m, resp, nil
+}
+
+// CreateTeamDiscussionCommentReaction creates a reaction for a team discussion comment
```
Comment (gmlewis) [inline github/reactions.go]: Please make this a complete sentence by adding a period at the end for the generated `godocs`.

File: github/reactions.go (hunk 1 above)
Comment (urohit011) [reply github/reactions.go]: Got it :+1: .

File: github/reactions.go (hunk 2 above)
Comment (urohit011) [reply github/reactions.go]: I'll do it. I see the comments of other endpoints method do not mention the possible reaction values so can I add them there as well?

File: github/reactions.go (hunk 5 above)
Comment (urohit011) [reply github/reactions.go]: Will do.

File: github/reactions.go (hunk 3 above)
Comment (urohit011) [reply github/reactions.go]: Will do.

File: github/reactions.go (hunk 2 above)
Comment (gmlewis) [reply github/reactions.go]: Sure! Thank you.

File: github/reactions.go (hunk 1 above)
Comment (juliaferraioli) [reply github/reactions.go]: Hmmm, I like your reasoning, Glenn. I think it's a nice subtle indication, so I concur.

Comment (gmlewis) [reply github/reactions.go]: OK, cool.  Thank you, @juliaferraioli!

Let's go ahead with the idea, and make these `*Number` fields `int`s.

**Comment by urohit011** [comment]
All requested changes made.

File: github/reactions_test.go [hunk 6]
```diff
@@ -184,6 +184,92 @@ func TestReactionsService_CreatePullRequestCommentReaction(t *testing.T) {
 	}
 }
 
+func TestReactionsService_ListTeamDiscussionReactions(t *testing.T) {
+	client, mux, _, teardown := setup()
+	defer teardown()
+
+	mux.HandleFunc("/teams/1/discussions/1/reactions", func(w http.ResponseWriter, r *http.Request) {
```
Comment (gmlewis) [inline github/reactions_test.go]: This is a tiny nit, but here and below, please make the parameters unique for two reasons:
* unique parameters make sure that the order and matching of name to value is correct, and
* it makes it easier to see which parameter goes where and affects which values for someone quickly glancing at the code.

So here and below, if you have two or more values that are the same (e.g. `1`), make then all unique.

**Comment by gmlewis** [comment]
@gauntface - Can you please have @willnorris add you as a contributor to this repo and I can start requesting reviews from you? Julia recommended that I contact you while she is unavailable for reviews.

Meanwhile, would you mind reviewing this PR for me and giving it an official LGTM if you approve?

Thank you!

**Comment by juliaferraioli** [comment]
@gmlewis @gauntface I've added Matt as a collaborator with read permissions as @willnorris is out.

File: github/reactions_test.go [hunk 7]
```diff
@@ -184,6 +184,92 @@ func TestReactionsService_CreatePullRequestCommentReaction(t *testing.T) {
...
+
+		w.WriteHeader(http.StatusOK)
+		w.Write([]byte(`[{"id":1,"user":{"login":"l","id":2},"content":"+1"}]`))
+	})
+
+	got, _, err := client.Reactions.ListTeamDiscussionReactions(context.Background(), 1, 2, nil)
+	if err != nil {
+		t.Errorf("ListTeamDiscussionReactions returned error: %v", err)
+	}
+	if want := []*Reaction{{ID: Int64(1), User: &User{Login: String("l"), ID: Int64(2)}, Content: String("+1")}}; !reflect.DeepEqual(got, want) {
```
[the remark this answers is not in the document]
Comment (juliaferraioli) [reply github/reactions_test.go]: Agreed, this makes it much cleaner (and consistent).

**Comment by urohit011** [comment]
Made the changes you suggested, @gauntface .

**Comment by gauntface** [comment]
Thank you so much @urohit011. Over to @gmlewis or @juliaferraioli to merge this PR :100:

**Comment by gmlewis** [comment]
Thanks, everyone!
Merging.

**Comment by urohit011** [comment]
Thanks everyone, really enjoyed working on this.

**Pull request merged**